snuba/state/rate_limit.py

```diff
@@ -0,0 +1,181 @@
+
+        try:
+            _, _, historical, concurrent = pipe.execute()
+            historical = int(historical)
+            concurrent = int(concurrent)
+            self.__did_run = True
+        except Exception as ex:
+            logger.exception(ex)
+            return {}  # fail open if redis is having issues
+
+        per_second = historical / float(state.rate_lookback_s)
+
+        rate_limit_name = self.__rate_limit_params.rate_limit_name
+
+        stats: Mapping[str, float] = {
+            '{}_rate'.format(rate_limit_name): per_second,
+            '{}_concurrent'.format(rate_limit_name): concurrent,
+
+        }
+
+        Reason = namedtuple('reason', 'scope name val limit')
+        reasons = [
+            Reason(rate_limit_name, 'concurrent', concurrent, self.__rate_limit_params.concurrent_limit),
+            Reason(rate_limit_name, 'per-second', per_second, self.__rate_limit_params.per_second_limit),
+        ]
+
+        reason = next((r for r in reasons if r.limit is not None and r.val > r.limit), None)
+
+        if reason:
+            self.__return_query_to_start_time()
+            raise RateLimitExceeded(
+                '{r.scope} {r.name} of {r.val:.0f} exceeds limit of {r.limit:.0f}'.format(r=reason)
+            )
+
+        return stats
+
+    def __return_query_to_start_time(self) -> None:
+        state.rds.zincrby(self.__bucket, self.__query_id, -float(state.max_query_duration_s))
+
+    def __exit__(
+            self,
+            exc_type: Optional[Type[BaseException]],
+            exc_val: Optional[BaseException],
+            exc_tb: Optional[TracebackType]
+    ) -> None:
+        try:
+            if self.__did_run:
+                self.__return_query_to_start_time()
+            else:
+                state.rds.zrem(self.__bucket, self.__query_id)  # not allowed / not counted
+        except Exception as ex:
+            logger.exception(ex)
+            pass
+
+
+def get_global_rate_limit_params() -> RateLimitParameters:
+    """
+    Returns the configuration object for the global rate limit
+    """
+
+    (per_second, concurr) = state.get_configs([
+        ('global_per_second_limit', None),
+        ('global_concurrent_limit', 1000),
+    ])
+
+    return RateLimitParameters(
+        rate_limit_name='global',
+        bucket='global',
+        per_second_limit=per_second,
+        concurrent_limit=concurr,
+    )
+
+
+class RateLimitAggregator(AbstractContextManager):
+    """
+    Runs the rate limits provided by the `rate_limit_params` configuration object.
+
+    It runs the rate limits in the order they are received and calls the cleanup block
+    (__exit__) in the reverse order.
+    """
+
+    def __init__(self, rate_limit_params: Sequence[RateLimitParameters]) -> None:
+        self.rate_limits = map(RateLimit, rate_limit_params)
+        self.stack = ExitStack()
+
+    def __enter__(self) -> MutableMapping[str, float]:
```

Comment:
I don't think you need to return a MutableMapping, the user of the context just need a Mapping. Even if under the hoods it is a mutable mapping at least the type checker would complain if the user tries to edit what it received.
Even better you can type alias it to give it a conceptual name so that who uses your context manager understand what that mapping represents.

Reply:
line 171 just returns it and to someone that should not mutate it. Are you sure about the line number ?
What I am suggesting is not to jsut use a Mapping, but to instantiate a MutableMapping but returning it as simple Mapping. MutableMapping is a subclass of Mapping, so, you can always tell the typechecker it is a Mapping and everyone receives it will think it is a Mapping, so the type checker will prevent mutations on it (of course you can downcast it to mutable mapping and mutate it, but then you deserve the dire consequences of mutability).